File: LegalDocParser/LegalDocParser/Helpers/AiSafetyFilter.cs
using System.Text.RegularExpressions;

namespace LegalDocParser.Helpers;

/// <summary>
/// AI 안전 필터: 프롬프트 인젝션 및 민감 정보 탐지
/// </summary>
public static class AiSafetyFilter
{
    private static readonly Regex ResidentIdPattern = new(
        @"\d{6}\s*-\s*[1-4]\d{6}",
        RegexOptions.Compiled);

    private static readonly Regex PhonePattern = new(
        @"01[0-9]-?\d{3,4}-?\d{4}",
        RegexOptions.Compiled);

    private static readonly Regex HiddenTextPattern = new(
        @"<\s*div[^>]*style\s*=\s*""[^""]*(?:display\s*:\s*none|visibility\s*:\s*hidden|font-size\s*:\s*0)[^""]*""[^>]*>.*?</\s*div\s*>",
        RegexOptions.Compiled | RegexOptions.IgnoreCase | RegexOptions.Singleline);

    private static readonly string[] PromptInjectionKeywords = new[]
    {
        "ignore previous instructions",
        "ignore above instructions",
        "disregard all prior",
        "system prompt",
        "you are now",
        "new instructions:",
        "override:",
        "이전 지시를 무시",
        "시스템 프롬프트"
    };

    public static (string filteredContent, List<string> warnings) Filter(string content)
    {
        var warnings = new List<string>();
        var filtered = content;

        // 숨겨진 텍스트 제거
        var hiddenMatches = HiddenTextPattern.Matches(filtered);
        if (hiddenMatches.Count > 0)
        {
            warnings.Add($"숨겨진 텍스트 {hiddenMatches.Count}개 발견 및 제거됨");
            filtered = HiddenTextPattern.Replace(filtered, "[HIDDEN_TEXT_REMOVED]");
        }

        // 프롬프트 인젝션 탐지
        foreach (var keyword in PromptInjectionKeywords)
        {
            if (filtered.Contains(keyword, StringComparison.OrdinalIgnoreCase))
            {
                warnings.Add($"프롬프트 인젝션 의심 패턴 발견: '{keyword}'");
            }
        }

        // 주민등록번호 탐지
        var residentMatches = ResidentIdPattern.Matches(filtered);
        if (residentMatches.Count > 0)
        {
            warnings.Add($"주민등록번호 패턴 {residentMatches.Count}개 발견");
        }

        // 전화번호 탐지
        var phoneMatches = PhonePattern.Matches(filtered);
        if (phoneMatches.Count > 0)
        {
            warnings.Add($"전화번호 패턴 {phoneMatches.Count}개 발견");
        }

        return (filtered, warnings);
    }

    public static string RedactSensitiveInfo(string content)
    {
        var result = ResidentIdPattern.Replace(content, "******-*******");
        result = PhonePattern.Replace(result, "***-****-****");
        return result;
    }
}

File: LegalDocParser/helpers/ai_safety_filter.py
import re


PROMPT_INJECTION_KEYWORDS = [
    "ignore previous instructions",
    "ignore above instructions",
    "disregard all prior",
    "system prompt",
    "you are now",
    "new instructions:",
    "override:",
    "이전 지시를 무시",
    "시스템 프롬프트",
]

RESIDENT_ID_PATTERN = re.compile(r"\d{6}\s*-\s*[1-4]\d{6}")
PHONE_PATTERN = re.compile(r"01[0-9]-?\d{3,4}-?\d{4}")
HIDDEN_HTML_PATTERN = re.compile(
    r"<div[^>]*(?:display\s*:\s*none|visibility\s*:\s*hidden|font-size\s*:\s*0)[^>]*>.*?</div>",
    re.DOTALL | re.IGNORECASE,
)


def filter_content(content: str) -> tuple[str, list[str]]:
    warnings: list[str] = []

    content_lower = content.lower()
    for keyword in PROMPT_INJECTION_KEYWORDS:
        if keyword.lower() in content_lower:
            warnings.append(f"프롬프트 인젝션 키워드 감지: '{keyword}'")

    if RESIDENT_ID_PATTERN.search(content):
        warnings.append("주민등록번호 패턴이 감지되었습니다.")

    if PHONE_PATTERN.search(content):
        warnings.append("전화번호 패턴이 감지되었습니다.")

    filtered = HIDDEN_HTML_PATTERN.sub("[HIDDEN_TEXT_REMOVED]", content)
    if filtered != content:
        warnings.append("숨겨진 HTML 텍스트가 제거되었습니다.")

    return filtered, warnings


def redact_sensitive_info(content: str) -> str:
    content = RESIDENT_ID_PATTERN.sub("******-*******", content)
    content = PHONE_PATTERN.sub("***-****-****", content)
    return content
